### PR TITLE
Set up server files and create user register endpoint #2

### LINKED ISSUE #2: Set up server files and create user register endpoint

### PR DESCRIPTION
**Pull request by enricodelarosa**

### CODE REVIEW

File: controllers/utils/validate.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    // If request body is blank
+    if (Object.keys(reqBody).length === 0 ) {
+        return {inputObj, isValid: false}
+    }
+
+    // Is valid true only if noen of the requeried fields is blank and the count of properties of req.body object is less than equal to the nubmer of needed keys
+    const isValid= validate.noneIsBlank(reqBody,neededKeys) && validate.countOfKeysMatch(reqBody, neededKeys);
+
+    return {reqBody, isValid}
+}
```
Comment (idiglove) [inline controllers/utils/validate.js]: im seeing a lot of misspellings and it should  be `isValid = `

File: routes/api/apiRoutes.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+const apiRouter = express.Router();
+
+const usersRouter = require('./users/usersRoutes.js');
+const gamesRouter = require('./games/gamesRouter.js');
+
+apiRouter.use('/users', usersRouter);
+
+apiRouter.use('/games', gamesRouter);
```
Comment (idiglove) [inline routes/api/apiRoutes.js]: shouldnt we have 

apiRouter.use('/reviews', reviewsRouter);

**Review by idiglove — COMMENTED** [review]
just a few things, and as discussed let's add TS support on it 👌🏽

File: routes/api/apiRoutes.js (hunk 2 above)
Comment (enricodelarosa) [reply routes/api/apiRoutes.js]: Oh right, we need a get all reviews endpoint. will and it by tomorrow

**Comment by enricodelarosa** [comment]
Already contains prisma schema for user, user_role, game, and review

File: prisma/schema.prisma [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+model user {
+  email      String    @id @db.VarChar(50)
+  username   String    @unique(map: "username") @db.VarChar(50)
+  first_name String    @db.VarChar(50)
+  last_name  String    @db.VarChar(50)
+  password   String    @db.Char(60)
+  role       String    @default("user") @db.VarChar(80)
+  user_role  user_role @relation(fields: [role], references: [id], onUpdate: Restrict, map: "user_ibfk_1")
```
Comment (idiglove) [inline prisma/schema.prisma]: what is the diff of role and user_role?

File: controllers/utils/validate.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    }
+    return true;
+}
+
+export const countOfKeysMatch = (Obj: any, neededKeys: any) => {
+
+    const countOfKeys = Object.keys(Obj).length;
+    return neededKeys.length <= countOfKeys; 
+
```
Comment (idiglove) [inline controllers/utils/validate.ts]: can we make sure that we don't use the type: `any` ?
Also, can we add a prettier/linter so that we don't have unnecessary whitespace and linebreaks?

File: controllers/usersController.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    res.send("hit get users route");
+}
+
+
+export const registerUser = async (req: any, res: any) => {
+
+    // I want to make this to a middleware? Or like a reusable function so that the isValid checking is not repeated and the reqBody properties would be available already as variables.
+    const neededKeys = ['email', 'username', 'firstName', 'lastName', 'password'];
+
+    const { reqBody, isValid } = validate.bodyVsKeys(req.body, neededKeys);
```
Comment (idiglove) [inline controllers/usersController.ts]: reading `validate.bodyVsKeys` doesnt make sense, can we rename it to something more understandable?

File: controllers/usersController.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+export const getAllUsers = (req: any, res: any) => {
+    res.send("hit get users route");
+}
+
+
+export const registerUser = async (req: any, res: any) => {
+
+    // I want to make this to a middleware? Or like a reusable function so that the isValid checking is not repeated and the reqBody properties would be available already as variables.
+    const neededKeys = ['email', 'username', 'firstName', 'lastName', 'password'];
```
Comment (idiglove) [inline controllers/usersController.ts]: yes you can do that
https://expressjs.com/en/guide/writing-middleware.html

File: controllers/utils/validate.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+export const bodyVsKeys = (reqBody: any, neededKeys: any) => {
+
+    // If request body is blank
+    if (Object.keys(reqBody).length === 0 ) {
+        return {reqBody, isValid: false}
+    }
+
+    // Is valid true only if noen of the requeried fields is blank and the count of properties of req.body object is less than equal to the nubmer of needed keys
+    const isValid = noneIsBlank(reqBody,neededKeys) && countOfKeysMatch(reqBody, neededKeys);
```
Comment (idiglove) [inline controllers/utils/validate.ts]: im seeing a lot of misspellings on the comments

Comment (idiglove) [reply controllers/utils/validate.ts]: like the engr manager from paymongo yesterday, he says that let's limit comments and instead use understandable code by making our function and variable names readable and easy to understand
in this case, I don't understand what noneIsBlank, and  countOfKeysMatch

File: prisma/schema.prisma (hunk 3 above)
Comment (enricodelarosa) [reply prisma/schema.prisma]: I'm actually not sure. What I did was create the user and user_role tables on mysql cli and use prisma db pull to generate the prisma schema. Will look into what line 18 means. But right now I think it establishes a one to many relation with the user_role table. Will look into what map means.

File: controllers/usersController.ts (hunk 5 above)
Comment (enricodelarosa) [reply controllers/usersController.ts]: Got it. Renaming now

File: prisma/schema.prisma (hunk 3 above)
Comment (enricodelarosa) [reply prisma/schema.prisma]: I looked it up in the prisma docs and i got.

onUpdate: Restrict - prevents the identifier of a referenced record from being changed.
In the context of the user_role and user tables, this means that user_roles id cannot be changed.

i can't find map: in the doccs, let me try to remove it and see what happens

Comment (enricodelarosa) [reply prisma/schema.prisma]: map: was using the user_ibkf_1 index for some reason. I suspect it's something that present on my tables since I created the tables initially using mysql cli and then did prisma db pull. I removed map in that relation and just used db push to make the tables according to the prisma schema.

**Comment by enricodelarosa** [comment]
I rewrote `validate.ts` so that the default export can be used as middleware.

Also changed the variable names so it's more readable. You're right, implying what the code does by using the variable and function names is pretty elegant! I hope I did it right haha.

Installed prettier and set `tabWidth` to 4 spaces.

Removed `any` types and replaced with more specific ones.

**Comment by enricodelarosa** [comment]
Added `outDir: dist` to tsconfig.json file.

**Comment by enricodelarosa** [comment]
Renamed files for consistency also added `"watch": true` to tsconfig.json

**Comment by enricodelarosa** [comment]
Renamed files for consistency also added `"watch": true` to tsconfig.json

**Review by EstopaceMA — CHANGES_REQUESTED** [review]
I think you should add a dist folder to be ignored in the gitignore file as it's not necessary to be committed since they're just locally generated build files.

**Comment by enricodelarosa** [comment]
Added dist folder to `.gitignore` file

File: prisma/schema.prisma (hunk 3 above)
Comment (idiglove) [reply prisma/schema.prisma]: Hi Rico, I don't think we have to worry about the map. How does the fields `role` and `user_role` relate to each other? I'm trying to understand hehe

File: controllers/utils/validate.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    requiredProps: string[]
+): returnObj => {
+    let objHasProps: boolean = true;
+    let missing: string[] = [];
+    let blank: string[] = [];
+
+    const isReqBodyEmpty = Object.keys(reqBody).length === 0;
+    if (isReqBodyEmpty) {
+        objHasProps = false;
+        let returnObj: returnObj = {
```
Comment (idiglove) [inline controllers/utils/validate.ts]: im loving the names :)
although i think we should refrain from using same variable names as with typenames. And types/interfaces convention initial char starts as capital

File: controllers/utils/validate.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    blank?: string;
+}
+
+export const checkObjHasProps = (
+    reqBody: object,
+    requiredProps: string[]
+): returnObj => {
+    let objHasProps: boolean = true;
+    let missing: string[] = [];
+    let blank: string[] = [];
```
Comment (idiglove) [inline controllers/utils/validate.ts]: hmm what is blank?

File: package.json [hunk 10]
```diff
@@ -0,0 +1,42 @@
+{
+    "name": "okunokumo.local",
```
Comment (idiglove) [inline package.json]: can we rename to okunokumo.be instead?

**Review by idiglove — CHANGES_REQUESTED** [review]

File: controllers/utils/validate.ts (hunk 8 above)
Comment (enricodelarosa) [reply controllers/utils/validate.ts]: Got it will be addressing this

File: controllers/utils/validate.ts (hunk 9 above)
Comment (enricodelarosa) [reply controllers/utils/validate.ts]: the props. It's defined (not missing) but has a blank value. Will change the property name to communicate that :)

File: package.json (hunk 10 above)
Comment (enricodelarosa) [reply package.json]: Got it

File: prisma/schema.prisma (hunk 3 above)
Comment (enricodelarosa) [reply prisma/schema.prisma]: Hi Fath, yes I did remove the map. A user can only have a role that is present in the user_role table. is that okay?

Comment (idiglove) [reply prisma/schema.prisma]: What is the use of user_role table?

Comment (enricodelarosa) [reply prisma/schema.prisma]: I wanted to set the role column of the user table to be a foreign key referencing the id column of the user_role table to establish a constraint. This would prevent the back end code from adding users with roles not present in the table. But now that I think about it, user_role table could actually be removed. Should I just remove user_role table?

roles available would be user, admin, superadmin and more can be added if necessary

**Review by idiglove — APPROVED** [review]
💯

**Review by EstopaceMA — APPROVED** [review]
LGTM! 💯
